src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/PhpFrameworkExtensionTest.php

```diff
@@ -82,4 +84,50 @@ public function testWorkflowValidationStateMachine()
             ]);
         });
     }
+
+    public function testRateLimiterWithLockFactory()
+    {
+        try {
+            $this->createContainerFromClosure(function (ContainerBuilder $container) {
+                $container->loadFromExtension('framework', [
+                    'rate_limiter' => [
+                        'with_lock' => ['policy' => 'fixed_window', 'limit' => 10, 'interval' => '1 hour'],
+                    ],
+                ]);
+            });
+
+            $this->fail('No LogicException thrown');
+        } catch (LogicException $e) {
+            $this->assertEquals('Rate limiter "with_lock" requires the Lock component to be installed and configured.', $e->getMessage());
+        }
+
+        $container = $this->createContainerFromClosure(function (ContainerBuilder $container) {
+            $container->loadFromExtension('framework', [
+                'lock' => true,
+                'rate_limiter' => [
+                    'with_lock' => ['policy' => 'fixed_window', 'limit' => 10, 'interval' => '1 hour'],
+                ],
+            ]);
+        });
+
+        $withLock = $container->getDefinition('limiter.with_lock');
+        $this->assertEquals('lock.factory', (string) $withLock->getArgument(2));
+    }
+
+    public function testRateLimiterLockFactory()
+    {
+        $this->markTestSkipped(
+        $this->expectException(OutOfBoundsException::class);
+        $this->expectExceptionMessage('The argument "2" doesn\'t exist.');
```

Comment:
Move these as far down as you can.

Reply:
May I ask you to elaborate why? Any occurrence I've come across in Symfony has these expect methods as first lines (see e.g. most of the tests above this in this file)

Reply:
Sure, Im sorry. 

We want to make sure the exception is throw when we execute `$container->getDefinition()`, right?

If `$this->createContainerFromClosure()` throws an `OutOfBoundsException` the test will still pass. In this specific case, we use a detailed exception and you verify the message so it is most likely fine. But consider this made up extreme example:

```php
public function testExceptionIsThrown()
{
    $this->expectException({RuntimeException::class);
    $foo = new Foo(); 
    $foo->add('bar');

    $service = new AcmeService();
    $service->handle($foo);
}
```

The purpose of this test is to verify that `$service->handle($foo);` throws an exception. However, what if `$foo->add('bar');` also threw a `\RuntimeException`? We would never know. We could actually have a bug in `Foo::add()` and a bug in `AcmeService::handle()` but the code looks like it is tested... 

The fix is to move down `exceptException()`. 

```diff
 public function testExceptionIsThrown()
 {
-    $this->expectException({RuntimeException::class);
     $foo = new Foo(); 
     $foo->add('bar'); 

     $service = new AcmeService();
+    $this->expectException({RuntimeException::class); 
     $service->handle($foo);
 }
```

Reply:
I see. Thanks for the detailed explanation! I don't necessarily agree 100% (because of the explicit class and message tests), but I've applied your suggestion